File: cpp/3904.smallest-stable-index-ii.cpp
#include <vector>
#include <utility>

using std::vector;

class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) {
        const int n = nums.size();
        
        int vmax = nums.front();
        vector<int> vmin(n, nums.back());

        for (int i = n - 2; i >= 0; i--) {
          vmin[i] = std::min(vmin[i + 1], nums[i]);
        }

        for (int i = 0; i < n; i++) {
          vmax = std::max(vmax, nums[i]);

          if (vmax - vmin[i] <= k) {
            return i;
          }
        }

        return -1;
    }
};

File: golang/3904.smallest-stable-index-ii.go
func firstStableIndex(nums []int, k int) int {
    n := len(nums)

    vmax := nums[0]
    
    vmin := make([]int, n)
    vmin[n - 1] = nums[n - 1]
    for i := n - 2; i >= 0; i-- {
      vmin[i] = min(vmin[i + 1], nums[i])
    }

    for i := 0; i < n; i++ {
      vmax = max(vmax, nums[i])

      if vmax - vmin[i] <= k {
        return i
      }
    }

    return -1
}